snuba/utils/streams/kafka.py

```diff
@@ -658,36 +658,30 @@ def __init__(
         self.__commit_log_topic = commit_log_topic
         self.__group_id = configuration["group.id"]
 
-    def poll(self, timeout: Optional[float] = None) -> Optional[Message[TPayload]]:
-        self.__producer.poll(0.0)
-        return super().poll(timeout)
-
     def __commit_message_delivery_callback(
-        self, error: Optional[KafkaError], message: ConfluentMessage
+        self, commit: Commit, future: Future[Commit]
     ) -> None:
-        if error is not None:
-            raise Exception(error.str())
+        try:
+            message = future.result()
+        except Exception:
+            logger.warning("Failed to produce commit record: %r", commit, exc_info=True)
+        else:
+            logger.debug("Produced commit record: %r", message)
 
     def commit_offsets(self) -> Mapping[Partition, int]:
         offsets = super().commit_offsets()
 
-        codec = CommitCodec()
         for partition, offset in offsets.items():
-            payload = codec.encode(Commit(self.__group_id, partition, offset))
-            self.__producer.produce(
-                self.__commit_log_topic.name,
-                key=payload.key,
-                value=payload.value,
-                on_delivery=self.__commit_message_delivery_callback,
+            commit = Commit(self.__group_id, partition, offset)
+            self.__producer.produce(self.__commit_log_topic, commit).add_done_callback(
+                partial(self.__commit_message_delivery_callback, commit)
             )
 
         return offsets
 
     def close(self, timeout: Optional[float] = None) -> None:
         super().close()
-        messages: int = self.__producer.flush(*[timeout] if timeout is not None else [])
-        if messages > 0:
-            raise TimeoutError(f"{messages} commit log messages pending delivery")
+        self.__producer.close().result(timeout)
```

Comment:
It drives me a bit crazy that this consumer is responsible for closing the producer. It seems as it should be the responsibility of the caller who provides it to the constructor, but making that change is more invasive than I'd like to get into now as it affects the stateful/snapshot consumer.

Reply:
Just remembered that the reason this drives me a bit crazy: this is because the consumer would now _require_ that this method is called to ensure a clean shutdown — including in exceptional cases, such as #762. 

So, this will not work any cases where [`ConsumerBuilder.build_base_consumer`](https://github.com/getsentry/snuba/blob/6c5a9c9f0630f6fcf08f5edd784a2aeaa55bdf9e/snuba/consumers/consumer_builder.py#L162-L175) is called without guaranteeing the returned consumer instance is closed.

The possible approaches here are:

1. ensure the consumer is closed (this is relatively straightforward for [the `consumer` command](https://github.com/getsentry/snuba/blob/6c5a9c9f0630f6fcf08f5edd784a2aeaa55bdf9e/snuba/cli/consumer.py#L148), less so for [the stateful consumer](https://github.com/getsentry/snuba/blob/6c5a9c9f0630f6fcf08f5edd784a2aeaa55bdf9e/snuba/stateful_consumer/states/consuming.py#L21).)
2. register an `atexit` handler to ensure that thread shutdown is requested if the main thread exits, similar to the approach in getsentry/sentry#15934
3. run the producer in a daemon thread and lose in-flight messages on exit without notice or warning
4. hope no errors ever occur and that the close method is always called by the signal handler and that everything is fine it is always fine nothing to see here
5. abandon ship

Reply:
Still going to need to figure this one out before it's ready.